clickhouse-admin/src/context.rs

```diff
@@ -3,84 +3,555 @@
+            .unwrap()
+            .join(CLICKHOUSE_SERVER_CONFIG_FILE);
+
+        // If there is already a configuration file with a generation number we'll
+        // use that. Otherwise, we set the generation number to None.
+        let gen = read_generation_from_file(config_path)?;
+        let (generation_tx, generation_rx) = watch::channel(gen);
+
+        // We only want to handle one in flight request at a time. Reconfigurator execution will retry
+        // again later anyway. We use flume bounded channels with a size of 0 to act as a rendezvous channel.
+        let (generate_config_tx, generate_config_rx) = flume::bounded(0);
+        tokio::spawn(long_running_generate_config_task(
+            generate_config_rx,
+            generation_tx,
+        ));
+
+        let (db_init_tx, db_init_rx) = flume::bounded(0);
+        tokio::spawn(long_running_db_init_task(db_init_rx));
+
+        Ok(Self {
             clickhouse_cli,
             clickward,
-            oximeter_client,
-            initialization_lock: Arc::new(Mutex::new(())),
+            clickhouse_address,
             log,
-        }
+            generate_config_tx,
+            generation_rx,
+            db_init_tx,
+        })
+    }
+
+    pub async fn send_generate_config_and_enable_svc(
+        &self,
+        replica_settings: ServerConfigurableSettings,
+    ) -> Result<GenerateConfigResult, HttpError> {
+        let clickward = self.clickward();
+        let log = self.log();
+        let node_settings = NodeSettings::Replica {
+            settings: replica_settings,
+            fmri: "svc:/oxide/clickhouse_server:default".to_string(),
+        };
+
+        let (response_tx, response_rx) = oneshot::channel();
+        self.generate_config_tx
+            .try_send(GenerateConfigRequest::GenerateConfig {
+                clickward,
+                log,
+                node_settings,
+                response: response_tx,
+            })
+            .map_err(|e| {
+                HttpError::for_internal_error(format!(
+                    "failure to send request: {e}"
+                ))
+            })?;
+        response_rx.await.map_err(|e| {
+            HttpError::for_internal_error(format!(
+                "failure to receive response: {e}"
+            ))
+        })?
+    }
+
+    pub async fn send_init_db(
+        &self,
+        replicated: bool,
+    ) -> Result<(), HttpError> {
+        let log = self.log();
+        let clickhouse_address = self.clickhouse_address();
+
+        let (response_tx, response_rx) = oneshot::channel();
+        self.db_init_tx
+            .try_send(DbInitRequest::DbInit {
+                clickhouse_address,
+                log,
+                replicated,
+                response: response_tx,
+            })
+            .map_err(|e| {
+                HttpError::for_internal_error(format!(
+                    "failure to send request: {e}"
```

Comment:
(Same notes as above about error handling and `{e}` probably not being very useful)